### PR TITLE
Out of range revisit and refactoring

### PR DESCRIPTION
**Pull request by xispa**
## Description of the issue/feature this PR addresses

This Pull Request contains several improvements/changes that raised up when trying to reduce the complexity of OutOfRange logic

### 1. Consistent `is_out_of_range` ###
Before this Pull Request, there were plenty of places in the code the out of range was computed differently and with several inconsistencies. In most cases, the use of adapters `IResultOutOfRange` was the preferred choice, especially because the criteria for determining if a result is out of range (and out of shoulders) depends on the type: `ReferenceAnalysis`, `DuplicateAnalysis` and `Analysis`. With this Pull Request, the function `api.analysis.is_out_of_range` has been created, all the old code that was calculating outofranges has been removed and this new function is used instead. To deal with the different criteria for different type of analysis types, the retrieval of the results ranges relies now in `getResultsRange` function that have been reimplemented especifically for each different type of analysis. Also, an effort to be sure all data returned from `getResultsRange` has been achieved.

### 2. Removal of % Error in Specifications ### 
% Error subfield is meaningless in result ranges. Also, the system was calculating the %Error based on the result (not based on the range), so the shoulder from the left was always smaller than the shoulder from the right. One could argue that this could be fixed easily by using the length of the valid range to compute the shoulders by using the %error. Again, this is a weak solution, cause in some cases, the scale of the range might be exponential, logaritmic (e.g. pH). Also, the sensitivity of the test may differ when the result falls in one end or another. Thus, is for the best interest of the labmanger to be able to set the shoulders (and therefore, when the warnings must show up) manually.

% Error hasn't been removed from `ReferenceSample`. Thus, this %error is still being used for `ReferenceAnalysis` - blank, control - (cause they don't need a valid range, rather a discrete result is expected).

### 3. Use % Error in ReferenceDefinition submission only ### 
% Error subfield is now only used on `ReferenceResults` submission (for `ReferenceSample` or `ReferenceDefinition`) to compute the `min` and `max` subfields, but is no longer stored in `ReferenceResults` schema field and not used anywhere else. This makes the whole thing more consistent, cause before this change, there was the chance to have a results range defined with values for min, max and result that were not consistent with the value set for % error.

### 4. Instrument validity not updated in accordance with latest QC tests  ###
Instrument's automatic validation based on QC Analyses (isQCValid) was not behaving properly. In some cases, the instrument was automatically set to "Valid" while an "in-range" result for a QC Analysis was submitted after a failing QC Analysis being submitted previously.

### 5. Inconsistencies when range was partially/fully negative ###
The system was not calculating shoulders and min/max (for reference definitions) when values (result, min or max) were negative numbers.

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: bika/lims/api/__init__.py [hunk 1]
```diff
@@ -1205,3 +1205,33 @@ def to_date(value, default=None):
...
+    :type value: str, float, int
+    :returns: The float number representation of the passed in value
+    :rtype: float
+    """
+    if not is_floatable(default):
+        fail("Param default needs to be floatable, got '%s'" % type(default))
+    try:
+        return float(value)
+    except:
+        return float(default)
```
Comment (ramonski) [inline bika/lims/api/__init__.py]: Make it more robust and return here `return to_float(default, default=0.0)`

File: bika/lims/api/__init__.py [hunk 2]
```diff
@@ -1205,3 +1205,33 @@ def to_date(value, default=None):
...
+    :param value: The value to be converted to a floatable number
+    :type value: str, float, int
+    :returns: The float number representation of the passed in value
+    :rtype: float
+    """
+    if not is_floatable(default):
+        fail("Param default needs to be floatable, got '%s'" % type(default))
+    try:
+        return float(value)
+    except:
```
Comment (ramonski) [inline bika/lims/api/__init__.py]: you are already testing for `is_floatable`, so you can remove the `try..except` here

File: bika/lims/api/__init__.py [hunk 3]
```diff
@@ -1205,3 +1205,33 @@ def to_date(value, default=None):
...
+
+
+def to_float(value, default=0.0):
+    """Converts the passed in value to a float number
+
+    :param value: The value to be converted to a floatable number
+    :type value: str, float, int
+    :returns: The float number representation of the passed in value
+    :rtype: float
+    """
```
Comment (ramonski) [inline bika/lims/api/__init__.py]: Please check out the module tests of this code:
https://github.com/senaite/senaite.jsonapi/blob/master/src/senaite/jsonapi/underscore.py

So maybe you can add some simple tests in these functions as well.

File: bika/lims/api/analysis.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    # Note that routine analyses, duplicates and reference analyses all them
+    # implement the function getResultRange:
+    # - For routine analyses, the function returns the valid range based on the
+    #   specs assigned during the creation process.
+    # - For duplicates, the valid range is the result of the analysis the
+    #   the duplicate was generated from +/- the duplicate variation.
+    # - For reference analyses, getResultRange returns the valid range as
+    #   indicated in the Reference Sample from which the analysis was created.
+    result_range = api.safe_getattr(analysis, "getResultsRange", None)
+    if not result_range or not isinstance(result_range, dict):
```
Comment (ramonski) [inline bika/lims/api/analysis.py]: This `isinstance(thing, dict)` is also something I see too often. Maybe we should add another util function `is_dict`?

File: bika/lims/api/__init__.py (hunk 3 above)
Comment (ramonski) [reply bika/lims/api/__init__.py]: Just saw that you have a doctest below, so you can ignore it or move your test here;)

File: bika/lims/api/analysis.py (hunk 4 above)
Comment (xispa) [reply bika/lims/api/analysis.py]: This is a temporary thing, cause atm, values returned by `getResultRange` by `Analysis`, `ReferenceAnalysis` and `DuplicateAnalysis` are not consistent. Will remove this `isinstance` as soon as I have this `getResultRange` standarized everywhere.

File: bika/lims/api/__init__.py (hunk 2 above)
Comment (xispa) [reply bika/lims/api/__init__.py]: I am testing `is_floatable` for `default` param, not for `value`, so is still necessary here.

File: bika/lims/content/referencesample.py [hunk 5]
```diff
@@ -237,42 +224,21 @@ def getResultsRangeDict(self):
...
+        results
+        :param only_uids: returns a list of uids or a list of objects
+        :return: list of uids or AnalysisService objects
+        """
+        uids = map(lambda range: range['uid'], self.getReferenceResults())
+        uids = [uid for uid in uids if api.is_uid(uid)]
+        if only_uids:
+            return uids
+        brains = api.search({'UID': uids}, 'uid_catalog')
+        return [api.get_object(brain) for brain in brains]
```
Comment (ramonski) [inline bika/lims/content/referencesample.py]:
```python
return map(api.get_object, brains)
```

File: bika/lims/tests/doctests/API_analysis.rst [hunk 6]
```diff
@@ -0,0 +1,502 @@
+API Analysis
+============
+
+Th> api_analysis provides single functions for single purposes especifically
```
Comment (ramonski) [inline bika/lims/tests/doctests/API_analysis.rst]: typo?

File: bika/lims/tests/doctests/API_analysis.rst [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+    ...     'Priority': '1',
+    ... }
+
+    >>> ar = create_analysisrequest(client, request, values, service_uids)
+    >>> success = doActionFor(ar, 'receive')
+
+Create a new Worksheet and add the analyses:
+
+    >>> worksheet = api.create(portal.worksheets, "Worksheet")
+    >>> analyses = [api.get_object(an) for an  in ar.getAnalyses()]
```
Comment (ramonski) [inline bika/lims/tests/doctests/API_analysis.rst]:
```python
>>> analyses = map(api.get_object, ar.getAnalyses()
```

File: bika/lims/tests/doctests/API_analysis.rst [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+
+    >>> au_analysis.setResult(5.6)
+    >>> is_out_of_range(au_analysis)
+    (True, True)
+
+
+Check if results for duplicates are out of range
+------------------------------------------------
+
+Get the first duplicate analysis that comes from from `Au`:
```
Comment (ramonski) [inline bika/lims/tests/doctests/API_analysis.rst]: "from from"

File: bika/lims/upgrade/v01_02_004.py [hunk 9]
```diff
@@ -31,6 +32,41 @@ def upgrade(tool):
...
+        for spec in specs_rr:
+            min = spec.get('min', '')
+            max = spec.get('max', '')
+            error = api.to_float(spec.get('error', '0'), 0)
+            if api.is_floatable(min) and api.is_floatable(max) and error > 0:
+                # Add min_warn and max_warn fields
+                min = api.to_float(min)
+                max = api.to_float(max)
+                warn_min = min - (abs(min) * (error/100.0))
+                warn_max = max + (abs(max) * (error/100.0))
```
Comment (ramonski) [inline bika/lims/upgrade/v01_02_004.py]: always take care when overwriting builtins. No issue here, just to keep in mind.

File: bika/lims/validators.py [hunk 10]
```diff
@@ -749,66 +751,96 @@ def __call__(self, value, *args, **kwargs):
...
-        mins = request.get('min', {})[0]
-        maxs = request.get('max', {})[0]
-        errors = request.get('error', {})[0]
-
-        # We run through the validator once per form submit, and check all
-        # values
-        # this value in request prevents running once per subfield value.
-        key = instance.id + fieldname
+        # This value in request prevents running once per subfield value.
+        key = '{}{}'.format(instance.id, fieldname)
```
Comment (ramonski) [inline bika/lims/validators.py]: better use `instance.getId()`

File: bika/lims/validators.py [hunk 11]
```diff
@@ -749,66 +751,96 @@ def __call__(self, value, *args, **kwargs):
...
-                          "and 100")))
-                return instance.REQUEST[key]
+        # For each parameter (e.g. 'min'), get the dict where the key is the
+        # uid of the Analysis Service and the value is the parameter value
+        mins = request.get('min', [{}])[0]
+        maxs = request.get('max', [{}])[0]
+        warn_mins = request.get('warn_min', [{}])[0]
+        warn_maxs = request.get('warn_max', [{}])[0]
+        errors = request.get('error', [{}])[0]
+        services = request.get('service', [{}])[0]
```
Comment (ramonski) [inline bika/lims/validators.py]: maybe a nicer syntax... 
```python
request.get('service') or {}
```

File: bika/lims/validators.py [hunk 12]
```diff
@@ -944,74 +976,61 @@ class ReferenceValuesValidator:
...
-                        _("Validation failed: Percentage error values must be "
-                          "between 0 and 100")))
+            err = errs.get(uid, None) or '0'
+            min = mins.get(uid, None) or res
+            max = maxs.get(uid, None) or res
+            if not api.is_floatable(res):
+                return to_utf8(translate(_("Validation for {} failed: "
+                                           "Expected result value must be "
+                                           "numeric".
+                                           format(service_name))))
```
Comment (ramonski) [inline bika/lims/validators.py]: not sure, but can `service_name` contain here non-ascii characters? Not that we get a `UnicodeDecodeError` when using `format` here...

File: bika/lims/browser/analysisrequest/templates/ar_add2.pt [hunk 13]
```diff
@@ -519,23 +519,29 @@
                                     tal:attributes="name string:${fieldname}.keyword:records;
                                                     value string:${service_keyword}"/>
                             <input type="text"
-                                   class="min" size="5"
+                                   class="warn_min" size="1"
```
[the remark this answers is not in the document]
Comment (xispa) [reply bika/lims/browser/analysisrequest/templates/ar_add2.pt]: [According to W3C](https://www.w3.org/TR/html401/interact/forms.html#adef-size-INPUT), `size` value refers to the (integer) number of characters to be displayed at once. Nevertheless, if there is neither `style:width=` attr nor css rule set, the browser automatically adjusts the width based on `size` value, that imo, is too wide. Given that this is `senaite.core` (and all UI styling will eventually end-up in `senaite.lims`), what do you think is better: add an 'style:width` attribute here in the template itself or add a rule in `ploneCustom.css.dtml`?

File: bika/lims/browser/analysisrequest/templates/ar_add2.pt [hunk 14]
```diff
@@ -696,23 +702,29 @@
...
                             <input type="text"
-                                   class="err" size="5"
-                                   placeholder="err%"
+                                   class="warn_max" size="1"
+                                   placeholder="&gt;max"
                                    i18n:attributes="placeholder"
-                                   tal:attributes="name string:${fieldname}.error:records;
-                                                   value python:service_spec.get('error');"/>
+                                   tal:attributes="name string:${fieldname}.warn_max:records;
+                                                   value python:service_spec.get('warn_max');"/>
```
[the remark this answers is not in the document]
Comment (xispa) [reply bika/lims/browser/analysisrequest/templates/ar_add2.pt]: That would be helpful, indeed. But that would be another PR

File: bika/lims/browser/chart/analyses.py [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+#
+# This file is part of SENAITE.CORE
+#
+# Copyright 2018 by it's authors.
+# Some rights reserved. See LICENSE.rst, CONTRIBUTORS.rst.
+
+import json
+from bika.lims import api
+
+class EvolutionChart():
```
Comment (ramonski) [inline bika/lims/browser/chart/analyses.py]: Please make this inherit from `object`

File: bika/lims/content/abstractroutineanalysis.py [hunk 16]
```diff
@@ -346,75 +346,29 @@ def getAnalysisRequestPrintStatus(self):
...
+        specs = {'min': '', 'max': '', 'warn_min': '', 'warn_max': ''}
+        analysis_request = self.getRequest()
+        if not analysis_request:
+            return specs
+
+        keyword = self.getKeyword()
+        ar_ranges = analysis_request.getResultsRange()
+        # Get the result range that corresponds to this specific analysis
+        an_range = [rr for rr in ar_ranges if rr.get('keyword', '') == keyword]
+        return an_range and an_range[0] or specs
```
[the remark this answers is not in the document]
Comment (xispa) [reply bika/lims/content/abstractroutineanalysis.py]: Did not recall on this one. Note the system is still constrained by the initial design, when there was the (bad) assumption that having an AR with more than one analysis from the same AS was a nonsense. Anyhow, you are right, unbinding keywords from results range and use UIDs instead could solve this problem. In another PR, promise!

**Pull request merged**